src/main/java/com/gradle/CustomBuildScanEnhancements.java

```diff
@@ -396,19 +397,20 @@ public void execute(BuildScanExtension buildScan) {
             }
 
             if (isNotEmpty(gitRepo) && isNotEmpty(gitCommitId)) {
-                if (gitRepo.contains("github.com/") || gitRepo.contains("github.com:")) {
-                    Matcher matcher = Pattern.compile("(.*)github\\.com[/|:](.*)").matcher(gitRepo);
-                    if (matcher.matches()) {
-                        String rawRepoPath = matcher.group(2);
-                        String repoPath = rawRepoPath.endsWith(".git") ? rawRepoPath.substring(0, rawRepoPath.length() - 4) : rawRepoPath;
-                        buildScan.link("Github source", "https://github.com/" + repoPath + "/tree/" + gitCommitId);
-                    }
-                } else if (gitRepo.contains("gitlab.com/") || gitRepo.contains("gitlab.com:")) {
-                    Matcher matcher = Pattern.compile("(.*)gitlab\\.com[/|:](.*)").matcher(gitRepo);
+                Optional<String> gitHubUrl = envVariable("GITHUB_SERVER_URL", providers);
+                Optional<String> gitHubRepository = envVariable("GITHUB_REPOSITORY", providers);
+                if (gitHubUrl.isPresent() && gitHubRepository.isPresent()) {
+                    buildScan.link("GitHub source", gitHubUrl + "/" + gitHubRepository + "/tree/" + gitCommitId);
+                } else {
+                    Matcher matcher = Pattern.compile("(?:https://|.*?@)(.*?(?:github|gitlab).*?)[:/](.*)\\.git").matcher(gitRepo);
```

Comment:
Why do you do all this pattern matching magic stuff instead of just creating a URI object?

Reply:
Updated to use `URI`. I agree that is the way to go. It is much clearer than the regex and also validates the URL returned by `git config --get remote.origin.url`.